### ISSUE TITLE
Missing documentation for `KeyedService.AnyKey`

### ISSUE THREAD

**Issue by aelij** [dotnet/docs#41346]
### Type of issue

Missing information

### Description

`KeyedService.AnyKey` can be used as a fallback when using keyed services.

### Page URL

https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection-guidelines

### Content source URL

https://github.com/dotnet/docs/blob/main/docs/core/extensions/dependency-injection-guidelines.md

### Document Version Independent Id

7fb2be80-cb54-27bf-4f93-423d8540429e

### Article author

@IEvangelist

### Metadata

* ID: 8d1c8ef1-6a01-df24-34ad-570f88df9303 
* Service: **dotnet-fundamentals**

---
[Associated WorkItem - 543725](https://dev.azure.com/msft-skilling/Content/_workitems/edit/543725)

**Comment by aelij** [comment]
@gewarren can you please also add that `GetKeyedServices<ISomething>(KeyedService.AnyKey)` can be used to get all keyed services?

**Comment by gewarren** [comment]
> [@gewarren](https://github.com/gewarren) can you please also add that `GetKeyedServices<ISomething>(KeyedService.AnyKey)` can be used to get all keyed services?

@aelij As of .NET 10, that behavior is no longer available. See https://learn.microsoft.com/en-us/dotnet/core/compatibility/extensions/10.0/getkeyedservice-anykey.

**Comment by aelij** [comment]
@gewarren 

What was removed changed in .NET 10 is if you used `AnyKey` to **add** the service it won't return in `GetKeyedServices`.

```cs
services.AddKeyedSingleton<IFoo, Foo>(KeyedService.AnyKey);
services.AddKeyedSingleton<IFoo, FooBar>("bar");
```

Before .NET 10 `GetKeyedServices<IFoo>` would return both `Foo` and `FooBar`, after - only `FooBar`.

This reminds me that an additional feature of `AnyKey` - using it for factories:

```cs
services.AddKeyedSingleton<ILogger>(KeyedService.AnyKey, (sp, key) => sp.GetRequiredService<ILoggerFactory>().CreateLogger(key.ToString());
```

Then we could inject `ILogger` with any key without registering it directly:

```cs
class Foo([FromKeyedServices("foo")] ILogger logger) { ... }
```

Can you please document both features?

**Comment by aelij** [comment]
And please reopen to keep track of this

**Comment by gewarren** [comment]
@ericstj Eli's description seems to contradict the breaking change announcement in https://github.com/dotnet/docs/issues/49885, which says that calling GetKeyedServices with AnyKey always returns an empty collection. Can you clarify?

**Comment by gewarren** [comment]
Based on @halter73 's [comment](https://github.com/dotnet/runtime/pull/113137#pullrequestreview-2680236499), it seems like the breaking change issue might not be correct about `GetKeyedServices`.

**Comment by aelij** [comment]
The breaking issue is correct - the wording might not be so clear 

> GetKeyedServices() (plural) no longer returns **AnyKey registrations** when queried with KeyedService.AnyKey

**Comment by ericstj** [comment]
Agreed with @aelij that this is what we said.  It's easy to miss the plural so maybe we can highlight that more in the doc.

**Comment by gewarren** [comment]
> Agreed with [@aelij](https://github.com/aelij) that this is what we said. It's easy to miss the plural so maybe we can highlight that more in the doc.

@ericstj The breaking change code snippet says that GetKeyedServices always returns an empty collection when queried with AnyKey, which I don't think is correct.

**Comment by ericstj** [comment]
Let me check where that came from -- bear with me, this was not my feature but one that I picked up and tried to document afterwards.

**Comment by ericstj** [comment]
This table better describes what is returned:
https://github.com/dotnet/runtime/pull/113137/changes#diff-5925c2c0e55020363d2025d55d6ca6a9dbec77e1b81d6bb99a9f687b98f6ce4bR41-R55

Query                     | Keyed? | Unkeyed? | AnyKey? | null key?
---------------------|---------|------------|-----------|----------
GetServices(Type)         | no     | yes      | no      | yes
GetService(Type)          | no     | yes      | no      | yes
GetKeyedServices(null)    | no     | yes      | no      | yes
GetKeyedService(null)     | no     | yes      | no      | yes
GetKeyedServices(AnyKey)  | yes    | no       | no      | no
GetKeyedService(AnyKey)   | throw  | throw    | throw   | throw
GetKeyedServices(key)     | yes    | no       | no      | no
GetKeyedService(key)      | yes    | no       | yes     | no
